src/de.rs

```diff
@@ -452,23 +498,101 @@ impl<S: Read> Decoder<S> {
             self.buffer().await?;
         }
 
-        if self.buffer.is_empty() {
-            Ok(())
-        } else {
-            if self.buffer.starts_with(QUOTE) {
-                self.parse_string().await?;
-            } else if self.numeric.contains(&self.buffer[0]) {
-                self.parse_number::<f64>().await?;
-            } else if self.buffer[0] == b'n' {
-                self.parse_unit().await?;
-            } else {
-                self.parse_bool().await?;
+        if !self.buffer.is_empty() {
+            // Determine the type of JSON value based on the first character in the buffer
+            match self.buffer[0] {
+                b'"' => self.ignore_string().await?,
+                b'-' => {
+                    self.eat_char().await?;
+                    self.ignore_number().await?;
+                }
+                b'0'..=b'9' => self.ignore_number().await?,
+                b't' => self.ignore_exactly("true").await?,
+                b'f' => self.ignore_exactly("false").await?,
+                b'n' => self.ignore_exactly("null").await?,
+                b'[' => self.ignore_array().await?,
+                b'{' => self.ignore_object().await?,
+                // If the first character doesn't match any JSON value type, return an error
+                _ => {
+                    return Err(Error::invalid_utf8(format!(
+                        "unexpected token ignoring value: {}",
+                        self.buffer[0]
+                    )))
+                }
             }
+        }
 
-            Ok(())
+        Ok(())
+    }
+
+    async fn ignore_string(&mut self) -> Result<(), Error> {
+        // eat the first char, which is a quote
+        self.next_or_eof().await?;
```

Comment:
```suggestion
        self.eat_char().await?;
```

Reply:
these functions basically do the same, it's just slightly clearer what's happening